src/tests/containerizer/cgroups2_tests.cpp

```diff
@@ -210,6 +221,119 @@ TEST_F(Cgroups2Test, ROOT_CGROUPS2_EnableAndDisable)
   EXPECT_EQ(0u, enabled->count("cpu"));
 }
 
+// Combination of a path and access flags.
+typedef pair<string, int> Access;
+
+using DeviceControllerTestParams = tuple<
+    vector<devices::Entry>, 
+    vector<devices::Entry>,
+    vector<Access>, 
+    vector<Access>>;
+
+class DeviceControllerTestFixture :
+    public Cgroups2Test,
+    public ::testing::WithParamInterface<DeviceControllerTestParams> {};
+
+
+TEST_P(DeviceControllerTestFixture, ROOT_CGROUPS2_DeviceController) {
+  const string& cgroup = TEST_CGROUP;
+
+  auto params = GetParam();
+  const vector<devices::Entry> allow = std::get<0>(params);
+  const vector<devices::Entry> deny = std::get<1>(params);
+  const vector<Access> allowedAccesses = std::get<2>(params);
+  const vector<Access> blockedAccesses = std::get<3>(params);
+
+  if (!cgroups2::exists(cgroup)) {
+    ASSERT_SOME(cgroups2::create(cgroup));
+  }
+  string path = cgroups2::path(cgroup);
+
+  ASSERT_SOME(devices::configure(cgroup, allow, deny));
+  Try<vector<uint32_t>> attached = ebpf::cgroups2::attached(path);
+  EXPECT_SOME(attached);
+  EXPECT_EQ(1u, attached->size());
+
+  pid_t pid = ::fork();
+  ASSERT_NE(-1, pid);
+
+  if (pid == 0) {
+    // Check that we can only do the "allowedAccesses".
+    foreach(const Access& access, allowedAccesses) {
+      ASSERT_SOME(os::open(access.first, access.second));
+    }
+    foreach(const Access& access, blockedAccesses) {
+      ASSERT_ERROR(os::open(access.first, access.second));
+    }
+
+    ASSERT_SOME(ebpf::cgroups2::detach(path, attached->at(0)));
+
+    // Check that we can do both the "allowedAccesses" and "blockedAccesses".
+    foreach(const Access& access, allowedAccesses) {
+      ASSERT_SOME(os::open(access.first, access.second));
+    }
+    foreach(const Access& access, blockedAccesses) {
+      ASSERT_SOME(os::open(access.first, access.second));
+    }
```

Comment:
We can't use ASSERT in the child process, as they won't fail the test. Looks like the strategy in cgroups_tests.cpp is to use a pipe and communicate back to the parent process to indicate whether the child process' tests succeeded.